Tidy invocation

tcorbettclark opened the pull request:
1. PEP8 coding style
2. Remove use of flask extensions because not appropriate/standard. It was only
   being used to provide a hook for authentication, which would probably be done
   separately by the main Flask app (or by a reverse proxy). Issue #73 requests basic auth which I will
   add in a separate pull request because it is unrelated to this tidy.
3. Removed app.py since it was not used and you would never use it from within a
   Flask app. The scripts/rq_dashboard.py provides an example use.
4. Emptied rq_dashboard/**init**.py for same reason as previous.
5. Moved fundamental dashboard configuration into separate file (one of
   recommended approaches).
6. Separated the command line configuration options (port and bind_addr) from
   the inherent rq dashboard options required for separate Flask hosting
7. Tidied how we set package version at build time and fetch version at run time.
8. Renamed dashboard.py to blueprint.py because we have too many "rq_dashboard"s
   and "dashboard"s around.

tcorbettclark committed:
Cleaned up dev-time and run-time retrieval of rq_dashboard version

tcorbettclark committed:
- PEP8 Coding style

- Remove use of flask extensions because not appropriate/standard. It was only
  being used to provide a hook for authentication, which would probably be done
  separately by the main Flask app (or by a reverse proxy).

- Removed app.py since it was not used and you would never use it from within a
  Flask app. The scripts/rq_dashboard.py provides an example use.

- Emptied rq_dashboard/__init__.py for same reason as previous.

- Moved fundamental dashboard configuration into separate file (one of
  recommended approaches).

- Separated the command line configuration options (port and bind_addr) from
  the inherent rq dashboard options required for separate Flask hosting

- Tidied how we set package version at build time and fetch version at run time.

- Renamed dashboard.py to blueprint.py because we have too many "rq_dashboard"s
  and "dashboard"s around.

tcorbettclark committed:
Clarified example use.

tcorbettclark committed:
Removed unnecessary quotes

tcorbettclark committed:
Typo

tcorbettclark committed:
Add HTTP Basic Auth for casual use.

tcorbettclark committed:
Fix #71

tcorbettclark committed:
Clarified docs and removed mistaken left over print

tcorbettclark committed:
Require PEP8 compliance before building

tcorbettclark committed:
Do not push tags as too dangerous, and remove leftover code

tcorbettclark committed:
Better opening docs

tcorbettclark committed:
Better docs

tcorbettclark committed:
Minor refactor of add_basic_auth mechanism (more orthogonal)

tcorbettclark committed:
Refactor order of Flask app+blueprint wiring (more logical grouping)

tcorbettclark committed:
Fixed reStructured text formatting typo

tcorbettclark committed:
Fixed reStructured text formatting typo (take 2)

tcorbettclark committed:
Tidy readme docs

joostdevries commented on setup.py:
```diff
@@ -1,14 +1,11 @@
-        exec(f.read(), locals)
-        return locals['VERSION']
-    raise RuntimeError('No version info found.')
+    """Return the latest git tag from repository in cwd."""
```
I wouldn't hardcode a git dependency in setup.py. I also think the code might not always be repo.

tcorbettclark replied:
Oh ok. What's the use-case that makes this a problem?

I use https://pypi.python.org/pypi/hgtools for my software in mercurial. It takes the version from the Hg tag. So this is a poor man's equivalent for git instead.

We unavoidably have version numbers in Git and the package itself (discoverable with pkg_resources from setuptools). We don't really want a third location. I've used files containing the version number before in the past, and it is so easy to forget to keep it in sync with the VC tag.

I suppose we could create a fab task to do the tagging. It would need to set the version in a file, add and check-in that file, then create the git tag. Seems fussy to me.

How strongly do you feel about it?

joostdevries replied:
Unless I'm wrong, `setup.py` runs during install. Since the packaged files are **not** a git repo and the users OS might not even have Git available, running `git` during install will result in errors. I think hgtools actually replaces the version getter during build.

tcorbettclark replied:
Ah, you are spot on. I pushed a test build to testpypi the other day and tested it with the following:

```
# Make a temp virtualenv
$ pew-mktmpenv

# The dependencies are not on testpypi so we have to pull them from the main pypi
$ pip instal rq flask times

# Install latest (but not yet released) rq-dashboard
$ pip install rq-dashboard -i https://testpypi.python.org/pypi

# Run it
$ rq-dashboard
RQ Dashboard version 0.3.5
 * Running on http://0.0.0.0:9181/ (Press CTRL+C to quit)
```

But this worked because we are using the wheel, which does not involve `setup.py` (one of it's features in fact). If instead we force use of the other package (which is a sdist) with:

```
$ pip install rq-dashboard -i https://testpypi.python.org/pypi --no-use-wheel
Collecting rq-dashboard
/home/tcorbettclark/.local/share/virtualenvs/9f4fc56de72d5ac/local/lib/python2.7/site-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:90: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. For more information, see https://urllib3.readthedocs.org/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
  Downloading https://testpypi.python.org/packages/source/r/rq-dashboard/rq-dashboard-0.3.5.tar.gz (92kB)
    100% |████████████████████████████████| 94kB 2.2MB/s 
    Complete output from command python setup.py egg_info:
    fatal: Not a git repository (or any of the parent directories): .git
    Traceback (most recent call last):
      File "<string>", line 20, in <module>
      File "/tmp/pip-build-CBen63/rq-dashboard/setup.py", line 13, in <module>
        version=get_version(),
      File "/tmp/pip-build-CBen63/rq-dashboard/setup.py", line 7, in get_version
        tags = subprocess.check_output(['git', 'tag']).strip().split('\n')
      File "/usr/lib/python2.7/subprocess.py", line 573, in check_output
        raise CalledProcessError(retcode, cmd, output=output)
    subprocess.CalledProcessError: Command '['git', 'tag']' returned non-zero exit status 128

    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-CBen63/rq-dashboard
```

tcorbettclark replied:
So: how about we use setuptools_git_version instead then?

It looks like it falls back to using the pkg_resource version if it has not been checked out in a git repo: https://github.com/pyfidelity/setuptools-git-version/blob/master/setuptools_git_version.py

I'm going to give it a quick go...

tcorbettclark replied:
That worked very nicely. I've updated the pull request. Thanks @joostdevries!

joostdevries commented on rq_dashboard/scripts/rq_dashboard.py:
```diff
@@ -71,17 +86,35 @@ def configure_app(app, options):
+    blueprint = rq_dashboard.blueprint.blueprint
+    options = get_options()
+    if options.username:
+        add_basic_auth(blueprint, options.username, options.password)
```
If we're going to provide basic auth, it seems that allowing these to be configured from ENV/config vars is reasonable.

tcorbettclark replied:
In other words, not just on the command line right? Hardly a good place for a password, even if it is only for casual use!

I was trying to keep separate the configuration of the blueprint (required however it is used) from the "configuration" of the `scripts/rq_dashboard.py` invocation. If we moved to http://click.pocoo.org/4/ as per #82 then retrieving settings from environment variables will be particular easy and give us that.

Does that sound like a plan to you?

joostdevries replied:
I'd either wait with introducing this feature or add config+ENV support now.

tcorbettclark replied:
Done by https://github.com/nvie/rq-dashboard/commit/df58151c74351195d8f4368e24ed137366b33749

username/password can now be passed in either on command line or through RQ_DASHBOARD_USERNAME/PASSWORD env var.

joostdevries commented on rq_dashboard/__init__.py:
```diff
@@ -1,19 +1,2 @@
-        app.register_blueprint(dashboard, url_prefix=self.url_prefix)
-        app.extensions['rq-dashboard'] = self
+import default_settings  # NOQA
+import blueprint  # NOQA
```
why not `from blueprint import blueprint`?

tcorbettclark replied:
Yes I agree that from e.g. scripts/rq_dashboard.py that would create a nicer API. But I felt it would be surprising and may even lead to circular import problems. Perhaps either the `blueprint.py` module or the `blueprint` variable should be named differently? Perhaps the module should be called `web.py`, especially if we also factor out the RQ status to (say) `rqstatus.py`. See https://github.com/nvie/rq-dashboard/issues/69#issuecomment-109722143.

Then the short cut in `__init__.py` as you suggest would work nicely.

What do you think?

joostdevries replied:
Even though we're in `0.x`, I'd avoid breaking API's wherever possible. If we're going to break it, let's do it right the first time. Doing `from rq.blueprint import blueprint` looks messy. Is there no nice way to make `from rq import blueprint` work?

tcorbettclark replied:
Fixed by https://github.com/nvie/rq-dashboard/commit/136cf65c3e53ea3fd5c13d795246a8c9677eec77

tcorbettclark committed:
Ignore .eggs/ directory

tcorbettclark committed:
Stop depending on git to run setup.py. See https://github.com/nvie/rq-dashboard/pull/80#discussion_r31875326

tcorbettclark committed:
Switched over to click for command line argument parsing. Fixes #82 and https://github.com/nvie/rq-dashboard/pull/80#discussion_r31890589

tcorbettclark committed:
Rename to avoid use of blueprint.blueprint in the Python API.

tcorbettclark committed:
Switched from times to arrow package. Fixes #81

tcorbettclark committed:
Typo and updated help docs

nvie committed:
Fix PEP8 complaint

nvie committed:
Add isort Fabric task

nvie committed:
Set file-wide flake8 ignoring

nvie committed:
Use isort to reorder import statements

nvie commented on setup.py:
```diff
@@ -25,10 +14,13 @@ def get_version():
     platforms='any',
+    # Manage version number from git at dev-time or pkg_resources at run-time.
+    setup_requires=['setuptools-git-version'],
+    version_format='{tag}',
```
Thanks for this pull request, @tcorbettclark! The only thing I don't like about it is the introduction of this install-time dependency on `setuptools-git-version`. It actually broke for me because of a dependency conflict with setuptools.

Suppose the latest version released to PyPI is 1.2.3. Then:
- On `master`, our WIP, version.py should report `1.2.4.dev0` to indicate a non-official version
- When ready, bump version.py to `1.2.4` (stable), tag it, release to PyPI, and change the version.py on `master` to `1.2.5.dev0`

This will also play well with `pip`'s rules for upgrading/downgrading.

I'll change this back to manual version tagging, then pull this PR. Thanks for the work!

tcorbettclark replied:
Great.

Yes, the hgtools version manages this better.

You're welcome. And @joostdevries was a great help too.

nvie committed:
Change back version detection to manual

This avoids the extra install-time dependency, and the need to be
installing this package from a Git repository.

nvie committed:
Indicate this is a WIP version

nvie committed:
Move {scripts/rq_dashboard => cli}.py
